File: Problem2_ReorderLinkedList.py
'''
====== Submission Details =======
Student Name: Pavan Kumar K. N.
S30 SlackID : RN32MAY2021
=================================
'''

# 143. Reorder List
# You are given the head of a singly linked-list. The list can be represented as:

# L0 → L1 → … → Ln - 1 → Ln
# Reorder the list to be on the following form:

# L0 → Ln → L1 → Ln - 1 → L2 → Ln - 2 → …
# You may not modify the values in the list's nodes. Only nodes themselves may 
# be changed.

#-----------------
# Time Complexity: 
#-----------------
# O(N) - N/2 operations to find mid. N/2 operations to reverse
#      - N/2 operations to merge reversed linked list
#------------------
# Space Complexity: 
#------------------
# O(1): Only temporary variables used. Constant space. 

#-----------------------
# Leet Code Performance: 
#-----------------------
# Ran Successfully?: Yes



# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reorderList(self, head: ListNode) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        #print(f"head:{head}")
        #print("Finding MID...")

        mid = head
        last = head
        while last and last.next:
            mid = mid.next
            last = last.next.next
        #print(f"MID:{mid.val}")
        
        curr = mid
        second_half = None
        
        #print(f"Second Half:{curr}")
        #print("Reversing...")
        
        while curr:
            temp = curr.next
            curr.next = second_half
            second_half = curr
            curr = temp
            
            
        #print(f"Reversed Second Half:{second_half}")
        
        first_curr = head
        second_curr = second_half
        while second_curr.next:
            temp = first_curr.next
            first_curr.next = second_curr
            first_curr = temp
            
            temp = second_curr.next
            second_curr.next = first_curr
            second_curr = temp
        
        #print(f"Head:{head}")

File: Problem2_ReorderLinkedList.cs
// Approach: Find the node previous to second half, and remove the link to second half to make two lists. Reverse second list and merge it with the first half
// Time Complexity: O(n). n/2 to find midpoint, n/2 to reverse the second half, n/2 to merge the lists.
// Space Complexity: O(1) No extra space is used other than some pointers.

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution {
    public void ReorderList(ListNode head) {
        ListNode slow = head, fast = head;
        
        while(fast.next!=null && fast.next.next!=null){
            fast = fast.next.next;
            slow = slow.next;
        }
        
        ListNode l2 = slow.next;
        slow.next = null;
        l2 = Reverse(l2);
        
        Merge(head, l2);
    }
    
    public ListNode Merge(ListNode l1, ListNode l2){
        ListNode preHead = new ListNode(-1, null);
        ListNode l3 = preHead;
        while(l1!=null && l2!=null){
            l3.next = l1;
            l1 = l1.next;
            l3 = l3.next;
            
            l3.next = l2;
            l2 = l2.next;
            l3 = l3.next;
        }
        
        l3.next = l1==null ? l2 : l1;
        
        return preHead.next;
    }
    
    public ListNode Reverse(ListNode head){
        ListNode prev = null;
        ListNode cur = head;
        
        while(cur!=null){
            ListNode next = cur.next;
            cur.next = prev;
            prev = cur;
            cur = next;
        }
        
        return prev;
    }
}